src/databricks/labs/ucx/azure/credentials.py

```diff
@@ -31,21 +33,35 @@ class ServicePrincipalMigrationInfo:
 
 @dataclass
 class StorageCredentialValidationResult:
-    name: str
+    name: str | None
     application_id: str
-    read_only: bool
+    read_only: bool | None
     validated_on: str
-    directory_id: str | None = None
+    directory_id: str | None = None  # str when storage credential created for AzureServicePrincipal
     failures: list[str] | None = None
 
     @classmethod
-    def from_validation(cls, permission_mapping: StoragePermissionMapping, failures: list[str] | None):
+    def from_storage_credential_info(
+        cls, storage_credential_info: StorageCredentialInfo, validated_on: str, failures: list[str] | None
+    ):
+        if storage_credential_info.azure_service_principal is not None:
+            application_id = storage_credential_info.azure_service_principal.application_id
+            directory_id = storage_credential_info.azure_service_principal.directory_id
+        elif storage_credential_info.azure_managed_identity is not None:
+            if storage_credential_info.azure_managed_identity.managed_identity_id is not None:
+                application_id = storage_credential_info.azure_managed_identity.managed_identity_id
+            else:
+                application_id = storage_credential_info.azure_managed_identity.access_connector_id
+            directory_id = None
+        else:
+            raise KeyError("Storage credential info is missing an application id.")
```

Comment:
i don't know if it's an overkill, but it _might get more clear_ if this is refactored to a separate private method
```suggestion
        if storage_credential_info.azure_service_principal is not None:
            application_id = storage_credential_info.azure_service_principal.application_id
            directory_id = storage_credential_info.azure_service_principal.directory_id
            return application_id, directory_id
        if storage_credential_info.azure_managed_identity is not None:
            if storage_credential_info.azure_managed_identity.managed_identity_id is not None:
                return storage_credential_info.azure_managed_identity.managed_identity_id, None
            return storage_credential_info.azure_managed_identity.access_connector_id, None
        raise KeyError("Storage credential info is missing an application id.")
```

Reply:
done